### PR TITLE
Deving branch

### PR DESCRIPTION
**Pull request by max-kov**
- migrated to numpy

- switched to pysprites

- refactored code

@kovalevvlad

### CODE REVIEW

**Comment by kovalevvlad** [comment]
Will take a look on Sunday :)

File: ball.py [hunk 1]
```diff
@@ -18,7 +18,7 @@ def __init__(self, ball_mass, planet_x, planet_y,is_striped,color,number,number_
 
         self.update_sprite()
 
-    def move_to(self, game_state, pos_x, pos_y):
+    def move_to(self, pos_x, pos_y):
```
Comment (kovalevvlad) [inline ball.py]: use vectors, be a pro :)

File: ball.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        pygame.sprite.Sprite.__init__(self)
+
+        self.mass = ball_data.ball_size
+        self.pos = np.array([0, 0], dtype=float)
+        self.velocity = np.array([0, 0], dtype=float)
+        self.radius = ball_data.ball_size
+        self.is_striped = ball_data.is_striped
+        self.color = ball_data.ball_color
+        self.number = ball_data.ball_number
+        self.number_text = ball_data.number_text
```
Comment (kovalevvlad) [inline ball.py]: number vs number_text?

File: ball.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        self.text_length = ball_data.number_text_size
+        self.sprite_offset = [0, 0]
+
+        self.update_sprite()
+
+    def move_to(self, pos_x, pos_y):
```
Comment (kovalevvlad) [inline ball.py]: use vectors

File: ball.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import numpy as np
+
+
+class Ball(pygame.sprite.Sprite):
+    def __init__(self, ball_data):
+        pygame.sprite.Sprite.__init__(self)
+
+        self.mass = ball_data.ball_size
+        self.pos = np.array([0, 0], dtype=float)
+        self.velocity = np.array([0, 0], dtype=float)
```
Comment (kovalevvlad) [inline ball.py]: https://docs.scipy.org/doc/numpy/reference/generated/numpy.zeros.html

File: ball.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        self.radius = ball_data.ball_size
+        self.is_striped = ball_data.is_striped
+        self.color = ball_data.ball_color
+        self.number = ball_data.ball_number
+        self.number_text = ball_data.number_text
+        self.text_length = ball_data.number_text_size
+        self.sprite_offset = [0, 0]
```
Comment (kovalevvlad) [inline ball.py]: use vectors - https://docs.scipy.org/doc/numpy/reference/generated/numpy.zeros.html

File: ball.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+        self.update_sprite()
+
+    def move_to(self, pos_x, pos_y):
+        self.pos = np.array([pos_x, pos_y], dtype=float)
+        self.rect.center = self.pos
+
+    def add_force(self, delta_x, delta_y):
+        self.velocity += [delta_x / self.mass, delta_y / self.mass]
```
Comment (kovalevvlad) [inline ball.py]: vectors please

File: ball.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        #     self.velocity[1] = 0
+        #
+        # if abs(self.velocity[0]) < 0.01:
+        #     self.velocity[0] = 0
+
+        self.top_left = self.pos - self.radius
+        self.rect.center = self.pos.tolist()
+
+    def set_vector(self, delta_x, delta_y):
+        self.velocity = np.array([delta_x, delta_y], dtype=float)
```
Comment (kovalevvlad) [inline ball.py]: vectors

File: ball.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    def update_sprite(self):
```
Comment (kovalevvlad) [inline ball.py]: let's discuss the whole method...

File: ballinfo.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import pygame
+import numpy as np
+
+# this class gives data about the individual balls
+# like the colour or the number on the ball
+
+class BallInfo:
```
Comment (kovalevvlad) [inline ballinfo.py]: how about just Ball?

File: ballinfo.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+            (200, 0, 200),
+            (200, 0, 0),
+            (50, 0, 0),
+            (100, 0, 0)
+        ]
+
+        self.ball_color = balls_colors[ball_number]
+        self.ball_size = ball_size
+        self.is_striped = ball_number > 8
+        self.ball_number = ball_number
```
Comment (kovalevvlad) [inline ballinfo.py]: if you change the class name to Ball then it makes sense to rename the fields:
```
ball_color -> color
ball_size -> size
```
etc.

File: collisiontests.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import physics
+import pygame
+
+
+def table_collision(game_state):
+    def ball_to_hole_collision(ball, hole):
+        return physics.point_distance(ball.pos, hole.pos) - hole.radius <= 0
```
Comment (kovalevvlad) [inline collisiontests.py]: this is already implemented in numpy (very likely)

Comment (kovalevvlad) [inline collisiontests.py]: http://stackoverflow.com/questions/1401712/how-can-the-euclidean-distance-be-calculated-with-numpy

File: collisiontests.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+            side.ball_hit(ball)
+
+
+def check_for_collision(game_state):
+    table_collision(game_state)
+
+    balls = game_state.balls
+    for counter1, ball1 in enumerate(balls):
+        for counter2, ball2 in enumerate(balls):
+            if not counter1 == counter2:
```
Comment (kovalevvlad) [inline collisiontests.py]: you can just compare ball1 != ball2 and get rid of the counters completely

File: cue.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+class Cue(pygame.sprite.Sprite):
+    def __init__(self, target, hit_power=0.8):
+        pygame.sprite.Sprite.__init__(self)
+
+        self.visible = True
+
+        self.target_ball = target
+
+        self.hit_power = hit_power
+        self.cue_length = 250
```
Comment (kovalevvlad) [inline cue.py]: get rid of cue_ suffixes, they don't make sense. Here is why. Imagine you have an instance:
```
cue = ....
cue.cue_length
```
wouldn't it be better to have 
```
cue.length?
```

File: cue.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+        sprite_centre = self.cue_length + self.max_displacement
+
+        self.image = pygame.Surface((2 * (self.cue_length + self.max_displacement),
+                                     2 * (self.cue_length + self.max_displacement)))
+        # color which will be ignored
+        self.image.fill((200, 200, 200))
+        self.image.set_colorkey((200, 200, 200))
+
+        cos_a = math.cos(math.radians(self.angle))
+        sin_a = math.sin(math.radians(self.angle))
```
Comment (kovalevvlad) [inline cue.py]: I think you can use vectors and matrices here

File: gamestate.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        self.all_sprites.add(self.sides)
+        self.all_sprites.add(self.holes)
+
+        self.canvas = graphics.Canvas(*self.resolution, background_color=self.table_color)
+
+        # fps control
+        self.fps_clock = pygame.time.Clock()
+        self.fps_limit = 100
+
+    def get_fps(self):
```
Comment (kovalevvlad) [inline gamestate.py]: just fps()

File: gamestate.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+            self.balls.add(ball.Ball(ballinfo.BallInfo(i)))
+
+    def set_pool_balls(self, x, y):
+        sin_60 = math.sin(math.radians(60))
+
+        diffx = sin_60 * self.ball_size * 2
+        diffy = 0.5 * self.ball_size * 4
+
+        ballx = 0
+        bally = 0
```
Comment (kovalevvlad) [inline gamestate.py]: vectors plz

File: gamestate.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+
+        for i, ball in enumerate(self.balls):
```
Comment (kovalevvlad) [inline gamestate.py]: you are not using i

File: graphics.py [hunk 18]
```diff
@@ -1,136 +1,87 @@
...
+        self.background = pygame.Surface(self.surface.get_size())
+        self.background = self.background.convert()
+        self.background.fill(background_color)
+
+        self.surface.blit(self.background, (0, 0))
+
+def draw_main_menu(game_state):
+    def check_mouse_pos(text_starting_place, text_ending_place, spacing, button_num):
+        mouse_pos = pygame.mouse.get_pos()
+        return (text_starting_place[button_num][0] - spacing < mouse_pos[0] < text_ending_place[button_num][
```
Comment (kovalevvlad) [inline graphics.py]: looks like you can probably vectorize this

File: ball.py (hunk 3 above)
Comment (max-kov) [inline ball.py]: vector = 2d numpy array?

Comment (kovalevvlad) [inline ball.py]: yez

**Pull request merged**